New factor levels in predict, round 2

michaellevy opened the pull request:
@mmastand this works now, and this is a smaller changeset as I've split it out into three separate PRs. Responses to your original review below. More verbose warnings will be added in #921, and the bug of `predict` sometimes breaking when an outcome column isn't provided will be addressed in #919. This closes #878.

- Warnings. There are no warnings when using predict and new data with new factor levels. They are silently imputed to common categories or means. Was that intentional? It looks like `verbose` is gone now, but I'd love to have the option to be told if new levels were coerced into other before generating predictions. I found it confusing to see that this dataframe produces the same prediction for all 3 rows but attaches the original data. Note that "Frozen" is a new level.    
    + Will be dealt with in #920

- If there is missingness in the original data, I get different results. I don't like handling things quietly. Null values in production could go completely unnoticed.  
    + I will add warnings in #920. Not sure what you about getting different results... if the original data is different, how could you get the same results?

- As a user, I'm not sure what `step_add_levels` is actually doing. It has the same behavior whether true or false. I can see in the recipe object that it's not getting used when FALSE, but those predictions shouldn't then be coming back the same, should they?
	+ It ensures that "other" and "hcai_missing" are added as levels to all nominal variables. If FALSE, they *may* be added if there are rare levels that are collapsed or missing values that are filled with `hcai_missing`; this *ensures* they are. Added a test that this is true and updated the docs to clarify this.

- If you prep and predict in 2 steps, you get the prepped data attached to the prediction rather than the original. Not necessarily bad, just different. I would guess that people usually wouldn't do this in favor of solely `predict`... 
    + The different output is intentional. If you let us do the prepping behind the curtain, as in `machine_learn` and `predict` without explicit prepping, you only ever see the original data. If you're more hands on and do the prepping yourself, you'll see the prepped data.
- ...Also can't pipe prep_data into predict because arguments are in different order. I could see people wanting to do this manually trying to do so...
    + `predict` is a generic function that always takes the model as the first argument; that's why its arguements are arranged this way. 
    + You can pipe `prep_data %>% predict`. Use `.` to place what's coming from the LHS of the pipe in an argument other than the first position:
```r
models <- machine_learn(pima_diabetes[1:50, ], diabetes)
pima_diabetes[51:70, ] %>%
  prep_data(recipe = models) %>%
  predict(models, .)
```
- ...Same output whether `prepdata=TRUE` or false.
    + `predict` tries to make a guess about whether prepping is needed; this switch is there just in case the user needs to override it. This is specified in the docs.

- With no missingness in my original data, categoricals are going to `other`, numerics are going to what?
    + numeric what? Missing values are handled by the numeric imputation method. Not sure I understand this question.

- I noticed that regression seems to fail when you don't give it a dummy outcome value. Separate issue, I would think.
    + Indeed. Now #922. In the future, feel free to create an issue with a minimal reproducible example when you find a bug.

- Prep data doesn't let new levels through, it seems to error because columns aren't lining up. That's probably due to not dummifying properly, which is the same as not letting new levels through 😄. Perhaps a better error though?
    + This works fine if you let `predict` figure out whether to prep the data or not. In this example, you're forcing it not to prep, but the data was prepped in training, so the format is off. I've tried to strengthen the language in the helpfile to direct the user to generally not use this.

- I fixed the `recipes:::printer` notes in my PR, thought it would be much better if Max exported.
    + Max has exported, but who knows when that will land on CRAN, so let's leave your solution for now.

Aylr commented on R/prep_data.R:
```diff
@@ -60,7 +60,9 @@
 #'   will be added to all nominal columns. This is protective in deployment: new
 #'   levels found in deployment will become "other" and missingness in
-#'   deployment can become "hcai_missing".
+#'   deployment can become "hcai_missing". If FALSE, these levels may be added
```
I don't like the user experience of "May" be added. Can I get more information as to why this appears non-deterministic from a user perspective?

michaellevy replied:
Add to docs; to answer your question: if the imputation method of categorical variables is to create a new level, then missing will become "hcai_missing".

Aylr commented on R/prep_data.R:
```diff
@@ -264,6 +266,11 @@ prep_data <- function(d,
     # If there are nominal predictors, apply nominal transformations
     if (any(var_info$type == "nominal" & var_info$role == "predictor")) {
+
+      # Add protective levels --------------------------------------------------
```
These kinds of comment headers are crufty and are code smells that refactoring needs to take place. We should consider breaking this 225 line function down and let the smaller functions *be the comment*.

Aylr replied:
That should happen in another PR that I'm happy to submit after this is merged.

michaellevy replied:
I'm all for it.

Aylr commented on tests/testthat/test-predict.R:
```diff
@@ -12,6 +12,12 @@ swiss <-
+te_new_missing$Catholic[6:7] <- NA
 # Need to test combos of:
 # reg/class x d from prep_data/not x newdata from prep_data/not
 ## Except training data not prepped and newdata prepped; that shouldn't work
```
Needing a decoder for variable names means they are probably terrible names.

michaellevy replied:
Totally true, and I hated working with this after I made it. Good call. Changed.

Aylr commented on tests/testthat/test-predict.R:
```diff
@@ -12,6 +12,12 @@ swiss <-
 part <- caret::createDataPartition(swiss$Catholic, .8)[[1]]
 tr <- swiss[part, ]
 te <- swiss[-part, ]
+te_newlevel <- te
```
These bits of code should be functions for reusability, clarity, and readability, and determinism. Having them in a file means that tests could interact with each other. Something like `make_missing_data()` or `make_missing_dataframe()`, which should be called every time an individual test runs.

michaellevy replied:
I like the idea of a having testing utility functions like `add_missingness(df, column)`. Going a step further, I like the idea of having a test-aaa.R file that creates one or two data frames that all the tests can be run against. We could do `prep_data`, `tune_models`, etc. there instead of across many test files and drop the time to run tests substantially. Added a note to #826.

Not sure I get the concern about tests interacting. We typically do setup at the top of test files for stuff that will be used across tests. Each `test_that` call is a self-contained environment.... they can read from the parent environment but can't (easily) modify objects there.

Aylr commented on tests/testthat/test-prep_data.R:
```diff
@@ -410,6 +410,12 @@ test_that("add_levels doesn't add levels to outcome", {
   expect_false(any(c("other", "hcai_missing") %in% levels(pd$x)))
 })
 
+test_that("prep_data respects add_levels = FALSE", {
```
See comments above on making test scaffolds generated by functions.

Aylr commented on tests/testthat/test-prep_data.R:
```diff
@@ -410,6 +410,12 @@ test_that("add_levels doesn't add levels to outcome", {
+test_that("prep_data respects add_levels = FALSE", {
+  d <- data.frame(x = c("A", "B"), y = 1:2)
+  pd <- prep_data(d, add_levels = FALSE, make_dummies = FALSE, impute = FALSE)
+  expect_false(any(c("other", "hcai_missing") %in% levels(pd$x)))
```
I may be doing something wrong, but I can't get this test to fail, and I suspect it never could.

The following snippet returns false, when I believe the intention for the test is that it should return true.

```
d <- data.frame(x = c("A", "B"), y = 1:2)
any(c("A")) %in% levels(d$x)
```

The fix should be a simple as moving a parenthesis around the `any` to the end of the line.

If this test were written first before writing the fix (TDD), it would have been immediately apparent that it was passing, and therefore a bad test.

michaellevy replied:
This test works, you're just testing the input data.frame (which does always have level "A" in "x") instead of the output data.frame. `any(c("A", "other") %in% levels(pd$x))` is TRUE.

michaellevy replied:
You also have the closing paren too early in your code snippet. In the test, `any` ends after `levels`.

michaellevy commented:
Great review, @Aylr. Let me make a few changes, let's punt on restructuring tests for now even though I think it's a great point (to do as part of #826), and then you can refactor to your heart's delight.

michaellevy commented:
@Aylr -- The workflow we've been using is that "request changes" means "I want to review your changes before merging"; just making comments means "Do these thing and then merge". I think I've addressed your concerns -- good to go? Excited to see how you'd refactor `prep_data` and get your hands dirty with R!
